File: Assembler c++/pass1/pass1.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class pass1 {
    public static void main(String[] args) throws Exception {
        FileReader fp = new FileReader("D:\\LP1-Practicals\\Assembler c++\\pass1\\input.txt");
        BufferedReader bufferReader = new BufferedReader(fp);

        int MAX = 101;

        // Pre-required tables for opcodes and directives
        Map<String, String> IsTab = new HashMap<>();
        Map<String, String> DsTab = new HashMap<>();
        Map<String, String> AdTab = new HashMap<>();

        // Initialize IsTab, DsTab, and AdTab
        // IMPERATIVE STATEMENTS
        IsTab.put("STOP", "00");
        IsTab.put("ADD", "01");
        IsTab.put("SUB", "02");
        IsTab.put("MULT", "03");
        IsTab.put("MOVER", "04");
        IsTab.put("MOVEM", "05");
        IsTab.put("COMP", "06");
        IsTab.put("BC", "07");
        IsTab.put("DIV", "08");
        IsTab.put("READ", "09");
        IsTab.put("PRINT", "10");

        // DECLARATIVE STATEMENTS
        DsTab.put("DS", "02");
        DsTab.put("DC", "01");

        // ASSEMBLER DIRECTIVES
        AdTab.put("ORIGIN", "03");
        AdTab.put("EQU", "04");
        AdTab.put("END", "02");
        AdTab.put("START", "01");
        AdTab.put("LTORG", "05");

        String[][] SymTab = new String[MAX][3];
        String[][] LitTab = new String[MAX][3];
        String[][] OpTab = new String[MAX][3];
        int[] PoolTab = new int[MAX];

        String line;
        int symline = 0, litline = 0, optabline = 0, poolIndex = 0;
        int LC = 0, linecnt = 0;

        // To store intermediate code (IC)
        List<String> intermediateCode = new ArrayList<>();

        System.out.println("--------CODE----------");

        while ((line = bufferReader.readLine()) != null) {
            String[] tokens = line.split("\t");

            // Print the current line
            System.out.println(line);

            if (tokens.length == 0) {
                System.out.println("Warning: Empty line encountered. Skipping.");
                continue;
            }

            if (linecnt == 0) {
               if (tokens.length > 1) {
                  LC = Integer.parseInt(tokens[1]);
               }
               linecnt++;
               continue;
            }

            // Process non-empty lines
            String label = tokens[0];
            String opcode = tokens.length > 1 ? tokens[1] : "";

            // Store Intermediate Code: Starting with the line counter
            StringBuilder icEntry = new StringBuilder(Integer.toString(LC));

            // Handle label
            if (!label.isEmpty()) {
                SymTab[symline][0] = label;
                SymTab[symline][1] = Integer.toString(LC);
                symline++;
            }

            // Handle opcode
            if (!opcode.isEmpty()) {
                icEntry.append("\t").append(opcode);

                // Handle Declarative Statements (DS, DC)
                if (opcode.equalsIgnoreCase("DS") || opcode.equalsIgnoreCase("DC")) {
                    if (tokens.length > 2) {
                        SymTab[symline - 1][2] = tokens[2];  // Store the value for DS/DC
                        icEntry.append("\t").append(tokens[2]);
                    }
                    LC++;
                }
                // Handle Literals
                else if (tokens.length > 2) {
                    String[] t2 = tokens[2].split(",");
                    if (t2.length > 1 && t2[1].charAt(0) == '=') {
                        LitTab[litline][0] = t2[1];
                        litline++;
                        icEntry.append("\t").append(t2[1]);
                    }
                }

                // Handle Assembler Directives
                if (opcode.equals("LTORG") || opcode.equals("END")) {
                    for (int i = poolIndex; i < litline; i++) {
                        if (LitTab[i][1] == null) {
                            LitTab[i][1] = Integer.toString(LC);
                            LC += 1;
                        }
                    }
                    PoolTab[poolIndex++] = litline;  // Update PoolTab
                } else if (opcode.equals("START")) {
                    if (tokens.length > 2) {
                        LC = Integer.parseInt(tokens[2]);
                        icEntry.append("\t").append(tokens[2]);
                    }
                } else if (opcode.equals("EQU")) {
                    String labelparam = tokens[2];
                    int rows = SymTab.length;
                    // Find the symbol in the symbol table and get its address
                    for (int i = 0; i < rows; i++) {
                        if (SymTab[i][0] != null && SymTab[i][0].equals(labelparam)) {
                            LC = Integer.parseInt(SymTab[i][1]);
                            break;
                        }
                    }
                    // Add the new label to the symbol table with the found address
                    SymTab[symline - 1][0] = label;
                    SymTab[symline - 1][1] = Integer.toString(LC);
                    symline++;
                } else if (opcode.equals("ORIGIN")) {
                    String expr = tokens[2];
                    if (expr.contains("+") || expr.contains("-")) {
                        String[] parts = expr.split("\\+|-");
                        String sym = parts[0].trim();
                        int value = Integer.parseInt(parts[1]);
                        for (int i = 0; i < symline; i++) {
                            if (SymTab[i][0].equals(sym)) {
                                LC = Integer.parseInt(SymTab[i][1]) + value;
                                break;
                            }
                        }
                    }
                } else if (IsTab.containsKey(opcode)) {
                    OpTab[optabline][0] = opcode;
                    OpTab[optabline][1] = "IS";
                    OpTab[optabline][2] = IsTab.get(opcode);
                    LC++;
                    optabline++;
                }
            }

            // Add the current IC entry to the list
            intermediateCode.add(icEntry.toString());
        }

        bufferReader.close();

        // Display Intermediate Code
        System.out.println("\n--------INTERMEDIATE CODE----------");
        for (String ic : intermediateCode) {
            System.out.println(ic);
        }

        // Display SymTab
        System.out.println("\n--------SYMBOL TABLE----------");
        System.out.println("Label\tAddress\tValue");
        for (int i = 0; i < symline; i++) {
            if (SymTab[i][0] != null) {
                System.out.println(SymTab[i][0] + "\t" + SymTab[i][1] + "\t" + (SymTab[i][2] != null ? SymTab[i][2] : ""));
            }
        }

        // Display LitTab
        System.out.println("\n--------LITERAL TABLE----------");
        System.out.println("Literal\tAddress");
        for (int i = 0; i < litline; i++) {
        	 if (LitTab[i][1] == null) {
        		 LitTab[i][1]=Integer.toString(LC);
             }
            if (LitTab[i][0] != null) {
                System.out.println(LitTab[i][0] + "\t" + LitTab[i][1]);
            }
        }

        // Display OpTab
        System.out.println("\n--------OPCODE TABLE----------");
        System.out.println("Opcode\tType\tValue");
        for (int i = 0; i < optabline; i++) {
            if (OpTab[i][0] != null) {
                System.out.println(OpTab[i][0] + "\t" + OpTab[i][1] + "\t" + OpTab[i][2]);
            }
        }
    }
}

File: Pass1.py
Symtab={}
Littab=[]
Pooltab=[0]
LC=0
IC=[]

IS={"STOP":0,"ADD":1,"SUB":2,"MUL":3,"MOVER":4,"MOVEM":5,"COMP":6,"BC":7,"DIV":8,"READ":9,"PRINT":10}
AD={"START":1,"END":2,"ORIGIN":3,"EQU":4,"LTORG":5}
DL={"DC":1,"DS":2}
REG={"AREG":1,"BREG":2,"CREG":3,"DREG":4}

def read_input_code(file_path):
    input_code=[]
    with open(file_path,'r') as file:
        for line in file:
            parts=[part.strip() for part in line.split(',')]
            input_code.append(parts)
    return input_code

Inputcode=read_input_code('input.txt')

def set_input(inputcode):
    global inputs
    inputs=inputcode

def pass1():
    global IC,LC
    for label,opcode,op1,op2 in inputs:
        if label and label!='-':
            Symtab[label]=LC
        if opcode in IS:
            IC.append(f"(IS,{IS[opcode]}) {get_opcode(op1)} {get_opcode(op2)}")
            LC+=1
        elif opcode in AD:
            IC.append(f"(AD,{AD[opcode]}) {get_opcode(op1)} {get_opcode(op2)}")
            handle_ad_directives(opcode,op1)
        elif opcode in DL:
            IC.append(f"(DL,{DL[opcode]}) {get_opcode(op1)}")
            LC+=1 if opcode=="DC" else int(op1)

def handle_ad_directives(opcode,op1):
    global LC
    if opcode=="START":
        LC=int(op1)
    elif opcode=="END" or opcode=="LTORG":
        processLiterals()
    elif opcode=="ORIGIN":
        LC=eval_expr(op1)
    elif opcode=="EQU":
        Symtab[op1]=eval_expr(op1)

def get_opcode(operand):
    if operand=="-":
        return ""
    elif operand in REG:
        return f"(R,{REG[operand]}"
    elif operand.startswith("="):
        Littab.append((operand,-1))
        return f"(L,{len(Littab)}"
    elif operand.isdigit():
        return f"(C,{operand})"
    else:
        return f"(S,{Symtab.get(operand,0)+1})"

def eval_expr(expr):
    return int(expr) if expr.isdigit() else Symtab.get(expr,0)

def processLiterals():
    global LC
    for i in range(Pooltab[-1],len(Littab)):
        Littab[i]=(Littab[i][0],LC)
        LC+=1
    Pooltab.append(len(Littab))

def show_tables():
    print("Intermediate Code")
    for code in IC:
        print(code)

    print("Symbol Table")
    for key,value in Symtab.items():
        print(f"{key}\t{value}")


    print("Literal Table")
    for literal in Littab:
        print(f"{literal[0]}\t{literal[1]}")


    print("Pool Table")
    for PoolIndex in Pooltab:
        print(PoolIndex)

def show_files():
    with open("ic.txt","w") as file:
        file.write("Intermediate Code\n")
        for code in IC:
            file.write(f"{code}\n")
    with open("st.txt","w") as file:
        for key,value in Symtab.items():
            file.write(f"{key}\t{value}\n")
    with open("lt.txt","w") as file:
        for literal in Littab:
            file.write(f"{literal[0]}\t{literal[1]}\n")
    with open("pt.txt","w") as file:
        for poolIndex in Pooltab:
            file.write(f"{poolIndex}\n")

set_input(Inputcode)
pass1()
show_tables()
show_files()






